### DISCUSSION TITLE
Granian http/https problem in Flask app

### DISCUSSION THREAD

**Discussion by kingkong-cmd** [emmett-framework/granian#475] [Q&A] [answered]
I just tried replacing gunicorn with granian for my flask app. Granian sits behind NGINX. Everything works fine, except all my links are now being resolved to http instead of https. I guess because granian sees the request as http://127.0.0.1:port. 

In gunicorn there is similar problem, but can be resolved by telling NGINX to add the headers below. Granian doesn't seem to treat these headers similarily as gunicorn. There is a possible fix in werkzung, but it's a bit messy, so was wondering if there is anyway to fix this on the granian/nginx level? 

```
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto https;
```

**Comment by gi0baro** [comment, marked as answer]
At current time, Granian won't use `X-Forwarded` headers to alter `wsgi.url_scheme` in WSGI environ, mainly 'cause there's no mention in WSGI spec that such value should be set using the mentioned headers. 
While we can discuss wether this behavior should be changed in future releases, at the time being this can probably be "fixed" by either:
- add SSL on Granian side and instruct nginx to communicate with the upstream via HTTPS
- wrap your WSGI application into an utility which sets `wsgi.url_scheme` to `https`, eg:

```python
def app_https_forw_wrapper(environ, start_response):
    if environ.get("HTTP_X_FORWARDED_PROTO") == "https":
        environ["wsgi.url_scheme"] = "https"
    return actual_app(environ, start_response)
```

**Reply by kingkong-cmd** [reply]
Ok, many thanks and understood on not in spec. 

I would like to avoid to deal with this on app level, since as a layman, I don't want to mess too much with stuff i don't really understand. Unless Flask guys rolls out something more integrated into settings/config.

For gunicorn it was just very convenient to address in config of nginx, so will probably just stay on gunicorn for now, and hope maybe there will be a clever option in the future for granian. Maybe it could simply be another settings parameter to enable proto (even if outside spec)? But I don't know, you're the expert.

Anyway, appreciate your work!

**Comment by edmorley** [comment]
As an additional data point, several other popular Python servers support this, e.g.:

Gunicorn (`--forwarded-allow-ips` and `FORWARDED_ALLOW_IPS`):
https://docs.gunicorn.org/en/stable/settings.html#secure-scheme-headers
https://docs.gunicorn.org/en/stable/settings.html#forwarded-allow-ips

uvicorn (`--forwarded-allow-ips` and `FORWARDED_ALLOW_IPS`):
https://www.uvicorn.org/deployment/#proxies-and-forwarded-headers
https://github.com/encode/uvicorn/blob/master/uvicorn/middleware/proxy_headers.py

Waitress (`--trusted-proxy` and `--trusted-proxy-headers`):
https://docs.pylonsproject.org/projects/waitress/en/latest/reverse-proxy.html

That said, some WSGI app frameworks/ecosystems also do support it at the app level too, e.g.:
- Django: https://docs.djangoproject.com/en/5.1/ref/settings/#secure-proxy-ssl-header
- flask-talisman: https://github.com/GoogleCloudPlatform/flask-talisman/blob/v0.7.0/flask_talisman/talisman.py#L211

**Comment by condemil** [comment]
Same issue with Starlette framework behind proxy.

As a suggestion, maybe it's possible make it configurable via cli arguments, like `--http-scheme [http|https]`.

**Reply by gi0baro** [reply]
See https://github.com/emmett-framework/granian/issues/389#issuecomment-2309834934
